src/Aspire.Hosting/Dashboard/DashboardAnnotation.cs

```diff
@@ -0,0 +1,93 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using Aspire.Dashboard.Model;
+using Aspire.Hosting.ApplicationModel;
+
+namespace Aspire.Hosting.Dashboard;
+
+/// <summary>
+/// The annotation that reflects how a resource shows up in the dashboard.
+/// </summary>
+public class DashboardAnnotation(Func<DashboardResourceState, DashboardResourceState> initialState) : IResourceAnnotation
+{
+    /// <summary>
+    /// The current state of the resource.
+    /// </summary>
+    public DashboardResourceState CurrentState { get; private set; } = new()
+    {
+        ResourceType = "",
+        Properties = []
+    };
+
+    /// <summary>
+    /// The callback that gets called when the state of the resource is updated.
+    /// </summary>
+    public Action<DashboardResourceState>? OnStateUpdated { get; set; }
+
+    /// <summary>
+    /// Creates the initial snapshot of the dashboard state for this resource.
+    /// </summary>
+    public DashboardResourceState IntializeState(DashboardResourceState state)
+    {
+        return CurrentState = initialState(state);
+    }
+
+    /// <summary>
+    /// Updates the snapshot of the dashboard state for a resource.
+    /// </summary>
+    /// <param name="state">The new <see cref="DashboardResourceState"/>.</param>
+    public void UpdateState(DashboardResourceState state)
+    {
+        OnStateUpdated?.Invoke(state);
+        CurrentState = state;
+    }
+}
+
+/// <summary>
+/// The context for a all of the properties and URLs that should show up in the dashboard for a resource.
+/// </summary>
+public record DashboardResourceState
+{
+    /// <summary>
+    /// The type of the resource.
+    /// </summary>
+    public required string ResourceType { get; init; }
+
+    /// <summary>
+    /// The properties that should show up in the dashboard for this resource.
+    /// </summary>
+    public required ImmutableArray<(string Key, string Value)> Properties { get; init; }
+
+    /// <summary>
+    /// Represents the state of the resource.
+    /// </summary>
+    public string? State { get; init; }
+
+    /// <summary>
+    /// The exit code of the resource.
+    /// </summary>
+    public int? ExitCode { get; init; }
```

Comment:
We should figure out of this is necessary on all resources.

Reply:
I'll remove it from here.